src/cluster.c

```diff
@@ -2255,10 +2260,11 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
             if (server.cluster->importing_slots_from[j]) continue;
 
             /* We rebind the slot to the new node claiming it if:
-             * 1) The slot was unassigned or the new node claims it with a
-             *    greater configEpoch.
+             * 1) The slot was unassigned or the previous owner no longer owns the slot or
+             *    the new node claims it with a greater configEpoch.
              * 2) We are not currently importing the slot. */
             if (server.cluster->slots[j] == NULL ||
+                bitmapTestBit(server.cluster->owner_not_owning_slot, j) ||
```

Comment:
I wonder if it is cleaner, more explicit, and less error-prone to abstract this check in a separate function or macro? There are multiple places where we now need to consider this new state.

In addition to the two places you updated, I think we need to patch `clusterSendFailoverAuthIfNeeded` as well in case one of the `node 1`'s replicas races ahead with a failover request.

There are also places where we check for full coverage, which I agree should not be affected.

Reply:
👍 I added the check in `clusterSendFailoverAuthIfNeeded`.

```
// clusterUpdateSlotsConfigWith

server.cluster->slots[j] == NULL ||
    bitmapTestBit(server.cluster->owner_not_owning_slot, j) ||
    server.cluster->slots[j]->configEpoch < senderConfigEpoch

// clusterProcessPacket

server.cluster->slots[j] == sender ||
    server.cluster->slots[j] == NULL ||
    bitmapTestBit(server.cluster->owner_not_owning_slot, j)

// clusterSendFailoverAuthIfNeeded

server.cluster->slots[j] == NULL ||
    bitmapTestBit(server.cluster->owner_not_owning_slot, j) ||
    server.cluster->slots[j]->configEpoch <= requestConfigEpoch
```

The common logic among these is checking if we a slot is bound with certainty. I tried using a macro for this

```
#define isSlotOwnershipSpeculative(slot) \
    (server.cluster->slots[slot] == NULL || \
        bitmapTestBit(server.cluster->owner_not_owning_slot, slot))
```